src/evaluator.ts

```diff
@@ -26,43 +33,111 @@ export class Evaluator {
+        addUnmatchedAllocation(AllocationEvaluationCode.BEFORE_START_TIME);
+        continue;
+      }
+      if (allocation.endAt && now > new Date(allocation.endAt)) {
+        addUnmatchedAllocation(AllocationEvaluationCode.AFTER_END_TIME);
+        continue;
+      }
+      const { matched, matchedRule } = matchesRules(
+        allocation?.rules ?? [],
+        { id: subjectKey, ...subjectAttributes },
+        obfuscated,
+      );
+      if (matched) {
         for (const split of allocation.splits) {
           if (
             split.shards.every((shard) => this.matchesShard(shard, subjectKey, flag.totalShards))
           ) {
+            const variation = flag.variations[split.variationKey];
+            const flagEvaluationDetails = flagEvaluationDetailsBuilder
+              .setMatch(i, variation, allocation, matchedRule, unmatchedAllocations)
+              .build(
+                'MATCH',
+                this.getMatchedEvaluationDetailsMessage(allocation, split, subjectKey),
+              );
             return {
               flagKey: flag.key,
               subjectKey,
               subjectAttributes,
               allocationKey: allocation.key,
-              variation: flag.variations[split.variationKey],
+              variation,
               extraLogging: split.extraLogging ?? {},
               doLog: allocation.doLog,
+              flagEvaluationDetails,
             };
           }
         }
+        // matched, but does not fall within split range
+        addUnmatchedAllocation(AllocationEvaluationCode.TRAFFIC_EXPOSURE_MISS);
+      } else {
+        addUnmatchedAllocation(AllocationEvaluationCode.FAILING_RULE);
       }
     }
-
-    return noneResult(flag.key, subjectKey, subjectAttributes);
+    return noneResult(
+      flag.key,
+      subjectKey,
+      subjectAttributes,
+      flagEvaluationDetailsBuilder
+        .setNoMatchFound(unmatchedAllocations)
+        .build(
+          'DEFAULT_ALLOCATION_NULL',
+          'No allocations matched. Falling back to "Default Allocation", serving NULL',
+        ),
+    );
   }
 
   matchesShard(shard: Shard, subjectKey: string, totalShards: number): boolean {
     const assignedShard = this.sharder.getShard(hashKey(shard.salt, subjectKey), totalShards);
     return shard.ranges.some((range) => isInShardRange(assignedShard, range));
   }
+
+  private getMatchedEvaluationDetailsMessage = (
+    allocation: Allocation,
+    split: Split,
+    subjectKey: string,
+  ): string => {
+    const hasDefinedRules = !!allocation.rules?.length;
+    const isExperiment = allocation.splits.length > 1;
+    const isPartialRollout = split.shards.length > 1;
+    const isExperimentOrPartialRollout = isExperiment || isPartialRollout;
+
+    if (hasDefinedRules && isExperimentOrPartialRollout) {
+      return `Supplied attributes match rules defined in allocation "${allocation.key}" and ${subjectKey} belongs to the range of traffic assigned to "${split.variationKey}".`;
```

Comment:
Is it possible to render the matched rule? Or is that tricky here?

Reply:
We do include the matched rule in the object returned. Do you mean to do it in the message too?

Reply:
Ah okay, nevermind then!